### DISCUSSION TITLE
zsh load is very slow

### DISCUSSION THREAD

**Discussion by brc-dd** [aws/amazon-q-developer-cli#202] [Q&A]
Same as https://github.com/aws/q-command-line-discussions/discussions/123 (seems to have regressed)

https://github.com/user-attachments/assets/c5677eb5-6724-43fa-963b-4b7c788e4c8b

```ini
[q-details]
version = "1.4.4"
hash = "956c5739077211bb2c5d4cb95d35c4ace3de42a0"
date = "2024-09-16T22:13:24.309363Z (9h ago)"

[system-info]
os = "macOS 14.7.0 (23H124)"
chip = "Apple M2 Pro"
total-cores = 10
memory = "16.00 GB"

[environment]
cwd = "/Users/USER/foo"
cli-path = "/Users/USER/.local/bin/q"
shell-path = "/bin/zsh"
shell-version = "5.9"
terminal = "VSCode"
install-method = "brew"

[env-vars]
PATH = "/Users/USER/.local/state/fnm_multishells/23575_1726556835035/bin:/opt/homebrew/bin:/opt/homebrew/sbin:/Users/USER/.composer/vendor/bin:/Users/USER/.deno/bin:/usr/local/bin:/System/Cryptexes/App/usr/bin:/usr/bin:/bin:/usr/sbin:/sbin:/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/local/bin:/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/bin:/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/appleinternal/bin:/Users/USER/.local/state/fnm_multishells/23501_1726556833848/bin:/opt/homebrew/bin:/opt/homebrew/sbin:/Users/USER/.composer/vendor/bin:/Users/USER/.deno/bin:/Users/USER/.local/bin:/Users/USER/Library/Application Support/JetBrains/Toolbox/scripts:/Users/USER/.dotnet/tools:/Users/USER/go/bin:/Users/USER/Library/Application Support/JetBrains/Toolbox/scripts:/Users/USER/.dotnet/tools:/Users/USER/go/bin"
QTERM_SESSION_ID = "84b0d31992b047c0a4d368fc1e66d686"
Q_SET_PARENT_CHECK = "1"
Q_TERM = "1.4.4"
SHELL = "/bin/zsh"
TERM = "xterm-256color"
ZDOTDIR = "/Users/USER"
__CFBundleIdentifier = "com.microsoft.VSCode"
```

**Comment by injust** [comment]
Maybe this isn't specific to zsh -- this looks like what happened to me on fish (for some time now).

**Reply by brc-dd** [reply]
Ah yeah. Some people on the original thread mentioned it happens on fish too.

> (for some time now)

Not sure about this part. Till around 1-2 weeks back it was fast enough for me.

**Comment by alpineyahoo** [comment]
My zsh is also slow. Do you see `.../brew.rb list amazon-q` process when zsh startup on your terminal? I guess this is the cause of slowness, but I don't know where this spec is defined and how to optimize it.

**Comment by ryanbrainard** [comment]
After removing all traces of Amazon Q CLI to try to improve slowness, it seemed to finally be the scripts it adds to`.zprofile` that did the trick.

I do like having Amazon Q in the CLI, but I don't need it always running or especially slowing down every time I open a new shell. I'd much rather have to wait when I actually need to use it. I understand the autocomplete stuff probably needs to be preloaded, but if you're just using the chat features, this could surely load on-demand, i would think.

**Comment by DarthGigi** [comment]
I'm having the same issue:

<details>
<summary>Video</summary>

https://github.com/user-attachments/assets/ea770d5f-b675-449b-9470-d4466d7d39ac

</details>

<details>
<summary>q diagnostic</summary>

```ini
[q-details]
version = "1.4.5"
hash = "804bfdf011c861a45b106573ced14fc088598781"
date = "2024-09-26T21:37:06.944937Z (14d ago)"

[system-info]
os = "macOS 15.0.1 (24A348)"
chip = "Intel(R) Core(TM) i7-9750H CPU @ 2.60GHz"
total-cores = 6
memory = "16.00 GB"

[environment]
cwd = "/Users/USER"
cli-path = "/Users/USER/.local/bin/q"
shell-path = "/bin/zsh"
shell-version = "5.9"
terminal = "iTerm 2"
install-method = "brew"

[env-vars]
PATH = "/usr/local/sbin:/Users/USER/Library/pnpm:/usr/local/opt/python/libexec/bin:/usr/local/opt/mysql@8.0/bin:/usr/local/opt/node@20/bin:/usr/local/bin:/System/Cryptexes/App/usr/bin:/usr/bin:/bin:/usr/sbin:/sbin:/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/local/bin:/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/bin:/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/appleinternal/bin:/Library/Apple/usr/bin:/Applications/Little Snitch.app/Contents/Components:/Applications/iTerm.app/Contents/Resources/utilities:/Users/USER/.local/bin"
QTERM_SESSION_ID = "8135ff08ec6e456f860a931c2edef8be"
Q_SET_PARENT_CHECK = "1"
Q_TERM = "1.4.5"
SHELL = "/bin/zsh"
TERM = "xterm-256color"
__CFBundleIdentifier = "com.googlecode.iterm2"
```
</details>

**Comment by peguimasid** [comment]
I've been dealing with this issue. After taking some time to investigate, I discovered that it’s related to enabling Q in VSCode. When I commented out the following line in vscode settings:

```jsonc
"terminal.integrated.env.osx": {
  // "Q_NEW_SESSION": "1"
},
```

the delay stopped occurring, but as a result, Amazon Q also stopped working.

**Comment by weierz** [comment]
Amazon Q collects and sends data to `eu-central-1-1.aws.cloud2.influxdata.com` on the start of every terminal session, even though I opt out auto update and the data collection in preference. Once I uninstalled amazon q, no more `curl` request has been made on the start and zsh loaded instantly.

<img width="848" alt="image" src="https://github.com/user-attachments/assets/50ee3abb-9185-411a-b9dc-4b8f6a4daec6">

**Comment by monsterhxw** [comment]
I have the same issue.

**Comment by sammcj** [comment]
Came to log this as well.

zsh load times:

- Without Amazon Q: 474ms
- With Amazon Q: 2137ms

**Comment by Vlaaaaaaad** [comment]
I too was annoyed by this and I moved to a manual Q enablement:
- deleted all related Q lines from `.zshrc` and `.zprofile` (deleting from just one file lowers the impact but does not completely eliminate it)
- added a new `manually_init_Q_because_AWS_is_customer_hostile()` function to my `.zshrc` that I manually call whenever I want to use Q (I was only interested in using `q chat` as I had and continue to have zero interest in Q taking over all my commands/shell):
```shell
manually_init_Q_because_AWS_is_customer_hostile() {
  # Amazon Q pre block for .zprofile
  [[ -f "${HOME}/Library/Application Support/amazon-q/shell/zprofile.pre.zsh" ]] && builtin source "${HOME}/Library/Application Support/amazon-q/shell/zprofile.pre.zsh"
  # Amazon Q post block for .zprofile
  [[ -f "${HOME}/Library/Application Support/amazon-q/shell/zprofile.post.zsh" ]] && builtin source "${HOME}/Library/Application Support/amazon-q/shell/zprofile.post.zsh"

  # Amazon Q pre block for .zshrc
  [[ -f "${HOME}/Library/Application Support/amazon-q/shell/zshrc.pre.zsh" ]] && builtin source "${HOME}/Library/Application Support/amazon-q/shell/zshrc.pre.zsh"
  # Amazon Q post block for .zshrc
  [[ -f "${HOME}/Library/Application Support/amazon-q/shell/zshrc.post.zsh" ]] && builtin source "${HOME}/Library/Application Support/amazon-q/shell/zshrc.post.zsh"
}
```
  

**Beware**: Amazon Q often re-enables itself and the Amazon Q app is full of dark patterns that will re-add those lines to `.zshrc` and `.zprofile`! For example, running `q doctor` shockingly does that because it _"auto-fixes things"_ 🤦

**Comment by ryanbrainard** [comment]
After dealing with this issue and uninstalling Amazon Q CLI, I thought I'd try it again, and I found that they now have a "Skip" option in the install wizard to avoid the take over of your shell. You have to wait a few seconds for it to show up and click it on a few of the steps. When you're done, `q` will be installed at `~/.local/bin/q` but without the rest of the crap. `q chat` still works this way even without @Vlaaaaaaad 's workaround.

**Reply by Vlaaaaaaad** [reply]
Oh, good to know `q chat` does not require any of the shell integration spam. Thank you for sharing that!

**Comment by sammcj** [comment]
I've conducted extensive analysis of the scripts that Amazon Q CLI injects into the shell, and have logged a bug report here: https://github.com/aws/amazon-q-developer-cli/issues/844

**Comment by GengCen-Qin** [comment]
I found a simple way to put Amazon Q directly into the Trash on Mac, which restores the original speed of the command line, but also makes Amazon Q inaccessible. When I need to use it again, I can drag Amazon Q back from the Trash to its original location, and it becomes usable again.  🤣

**Comment by sammcj** [comment]
Any update on this @aws? I completed extensive debugging for you several weeks back on https://github.com/aws/amazon-q-developer-cli/issues/844

**Comment by sammcj** [comment]
Any update on this @aws? Been quite a while and it's still slows down users terminal sessions to the point where they uninstall it.

**Comment by sammcj** [comment]
I haven't heard any updates from AWS, it appears they've lost interested in Amazon Q CLI 😞

**Comment by golimpio** [comment]
It's still very slow. I'm using fish and it's unusable, it takes several seconds to start the shell.

The `fig` integration they force us to use, even after checking out any cli completion isn't worth it. Fish has, out of the box, a faster completion than `fig` for the things I need.

`q chat` works well, but I gave up. Even manually uninstalling the shell integration (`fig`), `q` will re-install it in the background, so it's not like if I have a choice at all.

For some people (or most) I guess this isn't an issue, but I personally like to keep the starting time of my shell as fast as possible, since I use command line quite a lot.

As a good practice, they shouldn't force us to use `fig` as it's clearly buggy and slow at this stage.

When I say buggy, just to be fair, is because it required manually changing my zsh settings, as it didn't play well with my zinit setup. It's just my opinion, others may not consider this a bug.

**Comment by CampAsAChamp** [comment]
Still getting this as well despite https://github.com/aws/q-command-line-discussions/discussions/123 saying that it fixed it. My shell takes 2 seconds to load and be able to accept commands. I have created something similar to https://github.com/aws/amazon-q-developer-cli/discussions/202#discussioncomment-12515403 where I have a command to manually load the CLI scripts, but wish I didn't have to
